src/Compilers/CSharp/Test/Emit/CodeGen/PatternTests.cs

```diff
@@ -1125,6 +1125,198 @@ .maxstack  2
   IL_0013:  ret
   IL_0014:  ldc.i4.0
   IL_0015:  ret
+}");
+        }
+
+        [Fact, WorkItem(20641, "https://github.com/dotnet/roslyn/issues/20641")]
+        public void TupleSwitch01()
+        {
+            var source = @"using System;
+
+public class Door
+{
+    public DoorState State;
+
+    public enum DoorState { Opened, Closed, Locked }
+
+    public enum Action { Open, Close, Lock, Unlock }
+
+    public void Act(Action action, bool haveKey = false)
+    {
+        Console.Write($""{State} {action}{(haveKey ? "" withKey"" : null)}"");
+        State = ChangeState0(State, action, haveKey);
+        Console.WriteLine($"" -> {State}"");
+    }
+
+    public static DoorState ChangeState0(DoorState state, Action action, bool haveKey = false)
+    {
+        switch ((state, action))
+        {
+            case (DoorState.Opened, Action.Close):
+                return DoorState.Closed;
+            case (DoorState.Closed, Action.Open):
+                return DoorState.Opened;
+            case (DoorState.Closed, Action.Lock) when haveKey:
+                return DoorState.Locked;
+            case (DoorState.Locked, Action.Unlock) when haveKey:
+                return DoorState.Closed;
+            case var (oldState, _):
+                return oldState;
+        }
+    }
+
+    public static DoorState ChangeState1(DoorState state, Action action, bool haveKey = false) =>
+        (state, action) switch {
+            (DoorState.Opened, Action.Close) => DoorState.Closed,
+            (DoorState.Closed, Action.Open) => DoorState.Opened,
+            (DoorState.Closed, Action.Lock) when haveKey => DoorState.Locked,
+            (DoorState.Locked, Action.Unlock) when haveKey => DoorState.Closed,
+            _ => state };
+}
+
+class Program
+{
+    static void Main(string[] args)
+    {
+        var door = new Door();
+        door.Act(Door.Action.Close);
+        door.Act(Door.Action.Lock);
+        door.Act(Door.Action.Lock, true);
+        door.Act(Door.Action.Open);
+        door.Act(Door.Action.Unlock);
+        door.Act(Door.Action.Unlock, true);
+        door.Act(Door.Action.Open);
+    }
```

Comment:
>} [](start = 4, length = 1)

Consider executing with `ChangeState1` as well.

Reply:
Done already in a later iteration.

---
In reply to: [179658250](https://github.com/dotnet/roslyn/pull/25963#discussion_r179658250) [](ancestors = 179658250)